### ISSUE TITLE
MultiheadAttention forward shape of attn_mask 

### ISSUE THREAD

**Issue by badeaa3** [pytorch/pytorch#97605]
### 🐛 Describe the bug

Issue pertains to the argument `attn_mask` for nn.MultiheadAttention.

In train mode attn_mask is of shape (N⋅num_heads,L,L). 

In evaluation + no_grad mode when `attn_mask` and `key_padding_mask` are supplied then this line is triggered https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/activation.py#L1273. This expects an attn_mask of shape (L,L). This does not allow batched `attn_mask` and is incompatible with the train shape.

Example:
```
import torch
import torch.nn as nn

m = nn.MultiheadAttention(embed_dim=4, num_heads=4, batch_first=True) #Model()
N, L, E = 2, 8, 4
x = torch.randn(N,L,E)
key_padding_mask = torch.randn(N,L).bool()
attn_mask = torch.randn(N,L,L).repeat(m.num_heads,1,1)

m.train()
ao,aow = m(x,x,x,key_padding_mask=key_padding_mask,attn_mask=attn_mask)
print(ao.shape)

m.eval()
with torch.no_grad():
    try:
        ao,aow = m(x,x,x,key_padding_mask=key_padding_mask,attn_mask=attn_mask)
	print(ao.shape)
    except:
        print("doesn't work the same as training")
```

Is this a bug or if not what is the desired user input? 

A proposed fix would be to always require `attn_mask` input shape by user to be (N,L,L) and apply the expansion per head within the module.

### Versions

Collecting environment information...
PyTorch version: 2.0.0+cu117
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: CentOS Linux release 7.9.2009 (Core)  (x86_64)
GCC version: (GCC) 4.8.5 20150623 (Red Hat 4.8.5-44)
Clang version: Could not collect
CMake version: version 3.25.0
Libc version: glibc-2.17

Python version: 3.8.12 (default, Oct 12 2021, 13:49:34)  [GCC 7.5.0] (64-bit runtime)
Python platform: Linux-3.10.0-1160.88.1.el7.x86_64-x86_64-with-glibc2.17
Is CUDA available: True
CUDA runtime version: 11.3.109
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: Quadro RTX 4000
Nvidia driver version: 510.39.01
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                24
On-line CPU(s) list:   0-23
Thread(s) per core:    2
Core(s) per socket:    12
Socket(s):             1
NUMA node(s):          1
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 85
Model name:            Intel(R) Xeon(R) W-2265 CPU @ 3.50GHz
Stepping:              7
CPU MHz:               3355.804
CPU max MHz:           4800.0000
CPU min MHz:           1200.0000
BogoMIPS:              7000.00
Virtualization:        VT-x
L1d cache:             32K
L1i cache:             32K
L2 cache:              1024K
L3 cache:              19712K
NUMA node0 CPU(s):     0-23
Flags:                 fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc aperfmperf eagerfpu pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch epb cat_l3 cdp_l3 invpcid_single intel_ppin ssbd mba rsb_ctxsw ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid fsgsbase tsc_adjust bmi1 hle avx2 smep bmi2 erms invpcid rtm cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt clwb intel_pt avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req avx512_vnni md_clear spec_ctrl intel_stibp flush_l1d arch_capabilities

Versions of relevant libraries:
[pip3] numpy==1.24.2
[pip3] pytorch-lightning==2.0.0
[pip3] pytorch-triton==2.1.0+e650d3708b
[pip3] torch==2.0.0
[pip3] torchaudio==0.10.2
[pip3] torchmetrics==0.7.2
[pip3] torchvision==0.15.1
[pip3] triton==2.0.0
[conda] blas                      1.0                         mkl
[conda] cudatoolkit               11.3.1               h2bc3f7f_2
[conda] mkl                       2021.4.0           h06a4308_640
[conda] mkl-service               2.4.0            py38h7f8727e_0
[conda] mkl_fft                   1.3.1            py38hd3c417c_0
[conda] mkl_random                1.2.2            py38h51133e4_0
[conda] numpy                     1.24.2                   pypi_0    pypi
[conda] pytorch-lightning         2.0.0                    pypi_0    pypi
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] pytorch-triton            2.1.0+e650d3708b          pypi_0    pypi
[conda] torch                     2.0.0                    pypi_0    pypi
[conda] torchaudio                0.10.2               py38_cu113    pytorch
[conda] torchmetrics              0.7.2              pyhd8ed1ab_0    conda-forge
[conda] torchvision               0.15.1                   pypi_0    pypi
[conda] triton                    2.0.0                    pypi_0    pypi

**Comment by pat749** [comment]
**This is not a bug but rather a design choice made by PyTorch**. The behavior you are seeing is intended, as stated in the PyTorch documentation:
"In evaluation mode, batched multi-head attention requires a non-None key_padding_mask and a non-None attn_mask. attn_mask should be 2D with shape (L, L), where L is the sequence length, and is applied across all heads. key_padding_mask should be 2D with shape (N, L), where N is the batch size and L is the sequence length."

The reason for this design is that in evaluation mode, the multi-head attention layer computes attention scores across all heads for a single sequence of length L. Therefore, the attention mask only needs to be applied once to this sequence, rather than to each individual sequence in a batch. This is different from training mode, where the multi-head attention layer computes attention scores for each sequence in a batch separately, hence the need for a batched attention mask.
If you want to use the same attention mask for each sequence in a batch during evaluation, you can repeat the mask along the batch dimension before passing it to the multi-head attention layer. However, as you noted, this requires the attention mask to be of shape (N, L, L), which is different from the expected shape in evaluation mode. Therefore, your proposed fix of always requiring the attention mask input shape to be (N, L, L) and applying the expansion per head within the module would be a valid solution.

you can modify the **attn_mask** tensor before passing it to the **nn.MultiheadAttention** module during evaluation. Here's an updated version of the code:
```
import torch
import torch.nn as nn

m = nn.MultiheadAttention(embed_dim=4, num_heads=4, batch_first=True)
N, L, E = 2, 8, 4
x = torch.randn(N, L, E)
key_padding_mask = torch.randn(N, L).bool()
attn_mask = torch.randn(N, L, L).repeat(m.num_heads, 1, 1)

m.train()
ao, aow = m(x, x, x, key_padding_mask=key_padding_mask, attn_mask=attn_mask)
print(ao.shape)

m.eval()
with torch.no_grad():
    # Modify attn_mask to match expected shape
    attn_mask_eval = attn_mask.view(-1, L, L)
    try:
        ao, aow = m(x, x, x, key_padding_mask=key_padding_mask, attn_mask=attn_mask_eval)
        print(ao.shape)
    except:
        print("Error occurred during evaluation")
```
